### PR TITLE
fix-3619: use objectsync to construct msg object when objectsync uid …

### PR DESCRIPTION
**Pull request by neiba**
…not equal

Signed-off-by: chengengkun <1366542563@qq.com>

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes # https://github.com/kubeedge/kubeedge/issues/3619

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by vincentgoat — APPROVED** [review]

**Comment by vincentgoat** [comment]
cc @fisherxu @wackxu PTAL.

**Comment by neiba** [comment]
https://github.com/kubeedge/kubeedge/blob/6eaaf05fcd71ecbabd323403a35ad5b19eff955c/cloud/pkg/synccontroller/objectsync.go#L79-L81
If we use unstructured.Unstructured to init object, util.SetMetaType would error with: failed to set metatype: Object 'Kind' is missing in 'unstructured object has no kind'. Maybe we just need to set object uid ? @vincentgoat

**Comment by vincentgoat** [comment]
> https://github.com/kubeedge/kubeedge/blob/6eaaf05fcd71ecbabd323403a35ad5b19eff955c/cloud/pkg/synccontroller/objectsync.go#L79-L81
> 
> 
> If we use unstructured.Unstructured to init object, util.SetMetaType would error with: failed to set metatype: Object 'Kind' is missing in 'unstructured object has no kind'. Maybe we just need to set object uid ? @vincentgoat

agree

File: cloud/pkg/synccontroller/objectsync.go [hunk 1]
```diff
@@ -68,6 +68,7 @@ func (sctl *SyncController) manageObject(sync *v1alpha1.ObjectSync) {
 			Group:    "",
 			Resource: resource,
 		}, sync.Spec.ObjectName)
+		object.SetUID(types.UID(syncObjUID))
```
Comment (wackxu) [inline cloud/pkg/synccontroller/objectsync.go:71]: The change is tricky， it is strange and incomprehensible， we should update the object UID

Comment (wackxu) [reply cloud/pkg/synccontroller/objectsync.go:71]: why not do like this?
```
	newObject := &unstructured.Unstructured{}
		newObject.SetNamespace(sync.Namespace)
		newObject.SetName(sync.Spec.ObjectName)
		newObject.SetUID(types.UID(getObjectUID(sync.Name)))
		if msg := buildEdgeControllerMessage(nodeName, sync.Namespace, resourceType, sync.Spec.ObjectName, model.DeleteOperation, newObject); msg != nil {
			beehiveContext.Send(commonconst.DefaultContextSendModuleName, *msg)
		}
```

Comment (neiba) [reply cloud/pkg/synccontroller/objectsync.go:71]: Yes, i think this change is tricky too. If we send delete message and return here, then we are not need to construct NotFound error. Further more, sendEvents first argument is not need.

File: cloud/pkg/synccontroller/objectsync.go [hunk 2]
```diff
@@ -64,29 +52,37 @@ func (sctl *SyncController) manageObject(sync *v1alpha1.ObjectSync) {
...
+		sctl.deleteObjectSyncWithoutResource(sync)
+		return
 	}
 
-	sendEvents(err, nodeName, sync, resourceType, object.GetResourceVersion(), object)
+	sendEvents(nodeName, sync, resourceType, object.GetResourceVersion(), object)
 }
 
-func sendEvents(err error, nodeName string, sync *v1alpha1.ObjectSync, resourceType string,
+func (sctl *SyncController) deleteObjectSyncWithoutResource(sync *v1alpha1.ObjectSync) {
```
Comment (wackxu) [inline cloud/pkg/synccontroller/objectsync.go:62]:
```suggestion
func (sctl *SyncController) gcOrphanedObjectSync(sync *v1alpha1.ObjectSync) {
```

Comment (wackxu) [reply cloud/pkg/synccontroller/objectsync.go:62]: please also add comment to the function
```
// gcOrphanedObjectSync try to send delelte message to the edge node
// to make sure that the resource is deleted in the edge node. After the 
// message ACK is received by `cloudHub`, the objectSync will be deleted
// directly in the `cloudHub`. But if message build failed, the objectSync 
// will be deleted directly in the `syncController`.
```

Comment (neiba) [reply cloud/pkg/synccontroller/objectsync.go]: Updated, please take a look. @wackxu

**Review by vincentgoat — COMMENTED** [review]
lgtm, Thanks for your patience! @neiba

**Comment by wackxu** [comment]
releated to https://github.com/kubeedge/kubeedge/issues/3445

**Comment by wackxu** [comment]
we need merge this and some user has encountered this problem. cc @fisherxu @vincentgoat

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Comment by RyanZhaoXB** [comment]
Hello @wackxu , this PR may not resolve the problem in https://github.com/kubeedge/kubeedge/issues/3445. The problem still exists when user uses edgemesh for DNS resolving.

**Pull request merged**